### PR TITLE
New tool to check the usage of deprecated and unknown ESLint rules

### PR DESCRIPTION
**Pull request by frantic1048**
This PR resolves https://github.com/RightCapitalHQ/frontend-style-guide/issues/185

Limitation:

1. It currently only supports the legacy [`eslintrc`](https://eslint.org/docs/latest/use/configure/configuration-files-deprecated) config. We may extend support to [flat config](https://eslint.org/docs/latest/use/configure/configuration-files) in the future.
2. It currently only works with ESLint v8.

Misc:

1. Renamed `specs/eslint` to `specs/eslint-configs` for better semantics.
2. Used patched(merged, not released yet) fsWalk version that modified [packages/fs/fs.walk/src/providers/stream.ts ](https://github.com/nodelib/nodelib/pull/103/files#diff-15bc286e63f9d8f287dbfaf7757da16668648905f6d98ac82a904da40c3d6ade) to address the issue with fsWalk's stream interface.

### CODE REVIEW

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+const main = async () => {
+  const args = parseArgs({
+    options: {
+      cwd: {
+        type: 'string',
+        default: process.cwd(),
+      },
+      help: {
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:23]: `help` should go first

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+Usage: lint-eslint-config-rules [options]
+
+Options:
+  --cwd <path>\tThe directory to lint (default: process.cwd())
+  --json\tOutput all information in JSON format
+  -h, --help\tDisplay this help message
+
+Note:
+  "used" means the rule is specified in the config(including all extended configs), whether it's "off" or "warn" or "error".
+`);
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:59]: nit:

```suggestion
    console.log(`
${packageJson.name} v${packageJson.version}
${packageJson.description}

Usage: lint-eslint-config-rules [options]

Options:
  --cwd <path>\tThe directory to lint (default: process.cwd())
  --json\tOutput all information in JSON format
  -h, --help\tDisplay this help message

Note:
  "used" means the rule is specified in the config(including all extended configs), whether it's "off" or "warn" or "error".
`.trim());
```

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    return;
+  }
+
+  if (!json) {
+    console.log(`Checking ESLint rules in ${cwd}`);
+  }
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:65]: nit:

```suggestion
  console.error(`Checking ESLint rules in ${cwd}`);
```

File: packages/lint-eslint-config-rules/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  prefixedRuleId: string;
+  /** ruleId (without plugin prefix) */
+  ruleId: string;
+  meta?: Rule.RuleMetaData;
+}
+export type UsableRulesMap = Map<
+  /** equivalent to {@link UsableRule.prefixedRuleId} */
+  string,
+  IUsableRule
+>;
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:24]:
```suggestion
/** `${ruleId}` (from ESLint core) or `${pluginName}/${ruleName}` (from ESLint plugin) */
export type ESLintRuleId = string;
export interface IESLintRule {
  id: ESLintRuleId;
  meta: Rule.RuleMetaData | undefined;
}
export type ESLintRuleMap = Map<ESLintRuleId, IESLintRule>;
```

File: packages/lint-eslint-config-rules/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+export interface IESLintConfigRulesLintResult {
+  // other context
+  readonly eslint: ESLint;
+  readonly compat: FlatCompat;
+
+  readonly pluginsMap: Map<string, ESLint.Plugin>;
+  /** all rules that can be used */
+  readonly rulesMap: UsableRulesMap;
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:33]: nit:

```suggestion
  readonly pluginMap: Map<string, ESLint.Plugin>;
  /** all rules that can be used */
  readonly ruleMap: UsableRulesMap;
```

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    [key in keyof typeof args.values]: NonNullable<(typeof args.values)[key]>;
+  };
+
+  if (help) {
+    // MEMO: use JSON modules when it's stable to simplify this
+    // https://nodejs.org/api/esm.html#json-modules
+    const packageJson = JSON.parse(
+      await readFile(join(import.meta.dirname, '..', 'package.json'), {
+        encoding: 'utf8',
+      }),
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:45]:
```suggestion
      await readFile(import.meta.resolve('../package.json')),
```

File: packages/lint-eslint-config-rules/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+ */
+export const sortedRuleIds = (ruleIds: Iterable<string>): string[] =>
+  Array.from(ruleIds).sort((a, b) => {
+    if (a.includes('/') && !b.includes('/')) {
+      return 1;
+    }
+    if (!a.includes('/') && b.includes('/')) {
+      return -1;
+    }
+    return a.localeCompare(b);
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:62]: nit: `new Intl.Collator('en').compare(a, b)`

File: packages/lint-eslint-config-rules/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+   * Iterate over all files in the project to
+   * collect used rules and plugin specifiers
+   */
+  for (const entry of fsWalk.walkSync(
+    cwd,
+    new fsWalk.Settings({
+      deepFilter: (_entry) =>
+        !(
+          _entry.path.startsWith(join(cwd, 'node_modules')) ||
+          _entry.path.startsWith(join(cwd, '.git'))
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:96]: nit: 

```suggestion
          _entry.path.includes(`${path.sep}node_modules${path.sep}`) ||
          _entry.path.includes(`${path.sep}.git${path.sep}`)
```

**Review by lixiaoyan — CHANGES_REQUESTED** [review]

File: packages/lint-eslint-config-rules/src/index.ts (hunk 8 above)
Comment (frantic1048) [reply packages/lint-eslint-config-rules/src/index.ts]: Used `!['.git', 'node_modules'].includes(path.basename(_entry.path))` instead

File: packages/lint-eslint-config-rules/src/cli.ts (hunk 3 above)
Comment (frantic1048) [reply packages/lint-eslint-config-rules/src/cli.ts:67]: `console.error` will print red text in terminal, we may consider adding chalk to better format the output in the future 🤔

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  knownRuleIds: string[];
+  usedRuleIds: string[];
+  usedPluginSpecifiers: string[];
+  usedDeprecatedRuleIds: string[];
+  usedUnknownRuleIds: string[];
+}
+
+const main = async () => {
+  const args = parseArgs({
+    options: {
```
Comment (liby) [inline packages/lint-eslint-config-rules/src/cli.ts:17]: Consider offering option `--version`?

**Review by lixiaoyan — APPROVED** [review]

File: packages/lint-eslint-config-rules/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+      ruleId,
+      { id: ruleId, meta: rule.meta } satisfies IESLintRule,
+    ]),
+  );
+
+  /**
+   * Iterate over all files in the project to
+   * collect used rules and plugin specifiers
+   */
+  for (const entry of fsWalk.walkSync(
```
Comment (liby) [inline packages/lint-eslint-config-rules/src/index.ts:86]: Using `fsWalk.walkSync` with async `calculateConfigForFile` makes the loop sequential, impacting performance. Suggest refactoring to asynchronous `fsWalk.walk` with `Promise.all` for parallel execution.

```ts
const entries = await fsWalk.walk(cwd, settings);
const configs = await Promise.all(entries.map(entry => eslint.calculateConfigForFile(entry.path)));
```

File: specs/lint-eslint-config-rules/specs/eslint-8-eslintrc/.eslintrc.cjs
Comment (lixiaoyan) [inline specs/lint-eslint-config-rules/specs/eslint-8-eslintrc/.eslintrc.cjs:1]: specs -> fixtures

**Review by lixiaoyan — CHANGES_REQUESTED** [review]

File: packages/lint-eslint-config-rules/src/index.ts (hunk 10 above)
Comment (frantic1048) [reply packages/lint-eslint-config-rules/src/index.ts]: [`fsWalk.walk` is a callback-style API](https://www.npmjs.com/package/@nodelib/fs.walk#api), and it returns all entries at once, just like `fs.walkSync`. In my opinion, we cannot do much with it.

I've refactored the asynchronous `calculateConfigForFile` calls into a `Promise.all()`

Comment (liby) [reply packages/lint-eslint-config-rules/src/index.ts]: I apologize for the confusion in my previous comment. Upon further review of the fs.walk API documentation, I realized I overlooked an important detail. The documentation states: 
>  If you want to use the Promise API, use `util.promisify`.

Here's a corrected version that should work:

```ts
import { promisify } from 'util';

const walkAsync = promisify(fsWalk.walk);

const entries = await walkAsync(cwd, new fsWalk.Settings({
  deepFilter: (_entry) =>
    !['.git', 'node_modules'].includes(basename(_entry.path)),
  entryFilter: (_entry) => _entry.dirent.isFile(),
}));

await Promise.all(
  entries.map(async (entry) => {
    const config = await eslint.calculateConfigForFile(entry.path) as
      | undefined
      | {
          rules: Record<string, unknown>;
          plugins: string[];
        };

    if (config !== undefined) {
      usedRuleIds = usedRuleIds.union(new Set(Object.keys(config.rules)));
      usedPluginSpecifiers = usedPluginSpecifiers.union(
        new Set(config.plugins),
      );
    }
  })
);
```

This approach maintains the benefits of parallel processing that I initially suggested, while correctly handling the callback-style API of `fsWalk.walk`. Do you think this revised approach would be beneficial for the performance of the function?

File: packages/lint-eslint-config-rules/src/cli.ts (hunk 3 above)
Comment (liby) [reply packages/lint-eslint-config-rules/src/cli.ts:80]: Maybe we need to abstract a `logger`, and use package like [picocolors](https://www.npmjs.com/package/picocolors) again.

```
import { red  } from 'picocolors';

console.error(red('ERROR') + ' ', ...),
```

File: packages/lint-eslint-config-rules/src/index.ts (hunk 10 above)
Comment (frantic1048) [reply packages/lint-eslint-config-rules/src/index.ts]: Yes, we can use util.promisify() to convert fsWalk.walk to a promise-style function. However, we are still waiting all entries are returned before we can process them with `calculateConfigForFile`. The key steps are:

1. Get all entries from `fsWalk`, either by `await util.promisify(fsWalk.walk)(cwd)` or `fsWalk.walkSync(cwd)`.
2. Process entries from 1. with `calculateConfigForFile()`

The step 1 is a solid blocking operation, and we can't start step 2 until all entries are returned. Unless `fsWalk` provides a new API that returns its result as an async generator, we can't start processing each entry as soon as it is discovered.

The step 2 is a bunch of async operations, we can use `Promise.all()` to run them concurrently.

---

However, fsWalk provides a `fsWalk.walkStream()` API that returns results as a stream. We may use it to break down step 1 into smaller chunks, and process them concurrently.

Comment (frantic1048) [reply packages/lint-eslint-config-rules/src/index.ts]: The new implementation uses [`fsWalk.walkStream()`](https://github.com/nodelib/nodelib/tree/master/packages/fs/fs.walk#walkstreampath-optionsorsettingsi) and [`readable.forEach()`](https://nodejs.org/docs/latest/api/stream.html#readableforeachfn-options) to achieve fully asynchronous file enumeration and processing.

See the diff for more details:
https://github.com/RightCapitalHQ/frontend-style-guide/compare/5b5ab74381500aa74102fa4e2e9b0ca8870fa381..6f419a8f943894042fbc83acd8726a5b2bfecb69#diff-8c33b3645b2b6a257a0a0145134ed5295361df791d34eb6d62205be024ed2b2bL86-R123

File: patches/@nodelib__fs.walk@2.0.0.patch [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
++++ b/out/providers/stream.js
+@@ -26,10 +26,12 @@ class StreamProvider {
+         return new node_stream_1.Readable({
+             objectMode: true,
+             read: () => { },
+-            destroy: () => {
+-                if (!this.#reader.isDestroyed) {
+-                    this.#reader.destroy();
+-                }
++            destroy: (error, callback) => {
```
Comment (lixiaoyan) [inline patches/@nodelib__fs.walk@2.0.0.patch:13]: It will not work unless this patch is applied to each repository where the tool is used.

Comment (frantic1048) [reply patches/@nodelib__fs.walk@2.0.0.patch:13]: Oh, I forgot it. 😮  Let me find other workaround for this issue...

File: packages/lint-eslint-config-rules/package.json [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  },
+  "scripts": {
+    "prebuild": "pnpm run clean",
+    "build": "tsc --build",
+    "clean": "tsc --build --clean",
+    "prepack": "pnpm run build"
+  },
+  "dependencies": {
+    "@eslint/eslintrc": "3.1.0",
+    "@frantic1048/fs.walk": "2.0.1-pr103-async-iterator-fix",
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/package.json:33]:
```suggestion
    "@nodelib/fs.walk": "npm:@frantic1048/fs.walk@2.0.1-pr103-async-iterator-fix",
```

**Review by liby — APPROVED** [review]

**Review by lixiaoyan — APPROVED** [review]

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+Usage: lint-eslint-config-rules [options]
+
+Options:
+  -h, --help\tDisplay this help message
+  --cwd <path>\tThe directory to lint (default: process.cwd())
+  --json\tOutput all information in JSON format
+  --version\tDisplay version number
+
+Note:
+  "used" means the rule is specified in the config(including all extended configs), whether it's "off" or "warn" or "error".
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:72]:
```suggestion
  "used" means the rule is specified in the config (including all extended configs), whether it's "off" or "warn" or "error".
```

File: packages/lint-eslint-config-rules/src/cli.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        default: process.cwd(),
+      },
+      json: {
+        type: 'boolean',
+        default: false,
+      },
+    },
+  });
+
+  const { cwd, help, json, version } = args.values as {
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/cli.ts:39]:
```suggestion
  const { help, version, cwd, json } = args.values as {
```

File: packages/lint-eslint-config-rules/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+): Promise<IESLintConfigRulesLintResult> => {
+  const eslint = new ESLint({ cwd });
+  const compat = new FlatCompat({
+    baseDirectory: cwd,
+    resolvePluginsRelativeTo: cwd,
+  });
+
+  let usedRuleIds: Set<string> = new Set();
+  let usedPluginSpecifiers: Set<string> = new Set();
+  const rulesMap: ESLintRuleMap = new Map(
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:77]:
```suggestion
  const ruleMap: ESLintRuleMap = new Map(
```

File: packages/lint-eslint-config-rules/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+         * @see https://github.com/eslint/eslint/blob/63881dc11299aba1d0960747c199a4cf48d6b9c8/lib/eslint/eslint.js#L1208-L1212
+         */
+      },
+      {
+        concurrency: cpus().length * 2,
+      },
+    );
+
+  // resolve all plugins
+  const pluginsMap: Map<string, ESLint.Plugin> = new Map(
```
Comment (lixiaoyan) [inline packages/lint-eslint-config-rules/src/index.ts:126]:
```suggestion
  const pluginMap: Map<string, ESLint.Plugin> = new Map(
```

**Review by lixiaoyan — APPROVED** [review]

**Pull request merged**